File: godot-loader/src/nativeMain/kotlin/godot/loader/registry/bridge.kt
package godot.loader.registry

import godot.gdnative.godot_variant
import godot.loader.internal.NativeBindingContext
import godot.loader.internal.NativeKObject
import godot.loader.internal.NativeKVariant
import godot.loader.internal.nullSafe
import kotlinx.cinterop.*

fun createInstance(instance: COpaquePointer?, methodData: COpaquePointer?): COpaquePointer? {
    return NativeBindingContext.bindScope {
        val classHandle = nullSafe(methodData).asStableRef<NativeClassHandle>()
            .get()
        val kotlinInstance = classHandle.wrap(this, nullSafe(instance))
        kotlinInstance._onInit(this)
        val stableRef = StableRef.create(kotlinInstance)
        stableRef.asCPointer()
    }
}

fun disposeClassHandle(ref: COpaquePointer?) {
    NativeBindingContext.bindScope {
        val handle = nullSafe(ref).asStableRef<NativeClassHandle>()
        handle.get().dispose()
        handle.dispose()
    }
}

fun destroyInstance(instance: COpaquePointer?, methodData: COpaquePointer?, classData: COpaquePointer?) {
    NativeBindingContext.bindScope {
        val kotlinInstanceRef = nullSafe(classData).asStableRef<NativeKObject>()
        val kotlinInstance = kotlinInstanceRef.get()
        kotlinInstance._onDestroy(this)
        kotlinInstanceRef.dispose()
    }
}

fun invokeMethod(
    instance: COpaquePointer?,
    methodData: COpaquePointer?,
    classData: COpaquePointer?,
    numArgs: Int,
    args: CPointer<CPointerVar<godot_variant>>?
): CValue<godot_variant> {
    return NativeBindingContext.bindScope {
        val kotlinInstanceRef = nullSafe(classData).asStableRef<NativeKObject>()
        val kotlinInstance = kotlinInstanceRef.get()
        val funcHandleRef = nullSafe(methodData).asStableRef<NativeKFunc>()
        val funcHandle = funcHandleRef.get()

        val parameterCount  = funcHandle.getParameterCount(this)
        check(parameterCount == numArgs) {
            "Invalid number of arguments, $numArgs passed but $parameterCount expected."
        }

        val variantArgs = if (numArgs == 0) {
            emptyArray()
        } else {
            requireNotNull(args) { "args is null!" }
            Array(numArgs) { i -> NativeKVariant.fromGodot(this, args[i]!!.pointed.readValue())}
        }

        funcHandle(this, kotlinInstance, variantArgs).toGodot(this)
    }

}

File: godot-loader2/src/bridge.cpp
#include "bridge.h"
#include "native_binding_context.h"
#include <jvm.h>

void* createInstance(void* instance, void* methodData) {
    auto& bindingContext = NativeBindingContext::instance();
    bindingContext.startScope();
    auto& env = jni::Jvm::currentEnv();
    auto handle = (NativeClassHandle*) methodData;
    auto& classLoader = bindingContext.getClassLoader();
    auto kotlinInstance = handle->wrap(env, classLoader, instance);
    kotlinInstance->_onInit(env, classLoader);
    bindingContext.endScope();
    return kotlinInstance;
}

void disposeClassHandle(void* ref) {
    auto& bindingContext = NativeBindingContext::instance();
    bindingContext.startScope();
    auto& env = jni::Jvm::currentEnv();
    auto handle = (NativeClassHandle*) ref;
    handle->dispose(env);
    delete handle;
    bindingContext.endScope();
}

void destroyInstance(void* instance, void* methodData, void* classData) {
    auto& bindingContext = NativeBindingContext::instance();
    bindingContext.startScope();
    auto& env = jni::Jvm::currentEnv();
    auto kotlinInstance = (NativeKObject*) classData;
    auto& classLoader = bindingContext.getClassLoader();
    kotlinInstance->_onDestroy(env, classLoader);
    kotlinInstance->dispose(env);
    delete kotlinInstance;
    bindingContext.endScope();
}

void disposeFuncHandle(void* ref) {
    auto& bindingContext = NativeBindingContext::instance();
    bindingContext.startScope();
    auto& env = jni::Jvm::currentEnv();
    auto handle = (NativeKFunction*) ref;
    handle->dispose(env);
    delete handle;
    bindingContext.endScope();
}

godot_variant invokeMethod(void* instance, void* methodData, void* classData, int numArgs, godot_variant** args) {
    auto& bindingContext = NativeBindingContext::instance();
    bindingContext.startScope();
    auto& env = jni::Jvm::currentEnv();
    auto kotlinInstance = (NativeKObject*) classData;
    auto handle = (NativeKFunction*) methodData;

    auto parameterCount = handle->getParameterCount(env, bindingContext.getClassLoader());
    if (parameterCount != numArgs) {
        std::stringstream ss;
        ss << "Invalid number of args, expecting " << parameterCount << " but received " << numArgs;
        throw std::runtime_error(ss.str());
    }

    auto convertedArgs = std::vector<NativeTValue>();
    convertedArgs.reserve(numArgs);
    for (auto i = 0; i < numArgs; i++) {
        convertedArgs.emplace_back(NativeTValue(*args[i]));
    }

    auto res = handle->invoke(env, bindingContext.getClassLoader(), kotlinInstance, convertedArgs).toGVariant();
    bindingContext.endScope();
    return res;
}